File: android/app/src/main/java/com/pvarki/rasenmaeher/KeyChainModule.java
package com.yourapp;

import android.content.Context;
import android.security.KeyChain;
import android.security.KeyChainAliasCallback;
import android.util.Base64;
import android.util.Log;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import java.security.KeyStore;
import java.security.PrivateKey;
import java.security.cert.Certificate;
import java.security.cert.CertificateFactory;
import java.security.cert.X509Certificate;
import java.util.ArrayList;
import java.util.List;

public class KeyChainModule extends ReactContextBaseJavaModule {

    private static final String TAG = "KeyChainModule";
    private static ReactApplicationContext reactContext;

    KeyChainModule(ReactApplicationContext context) {
        super(context);
        reactContext = context;
    }

    @Override
    public String getName() {
        return "KeyChainModule";
    }

    @ReactMethod
    public void installCertificate(String p12, String name, Promise promise) {
        try {
            byte[] p12Bytes = Base64.decode(p12, Base64.DEFAULT);
            KeyStore keyStore = KeyStore.getInstance("PKCS12");
            keyStore.load(new java.io.ByteArrayInputStream(p12Bytes), "".toCharArray());

            String alias = keyStore.aliases().nextElement();
            KeyStore.PrivateKeyEntry entry = (KeyStore.PrivateKeyEntry) keyStore.getEntry(alias, null);

            PrivateKey privateKey = entry.getPrivateKey();
            Certificate[] certChain = entry.getCertificateChain();
            List<X509Certificate> certificates = new ArrayList<>();
            for (Certificate certificate : certChain) {
                certificates.add((X509Certificate) certificate);
            }

            KeyChain.choosePrivateKeyAlias(
                getCurrentActivity(),
                new KeyChainAliasCallback() {
                    @Override
                    public void alias(String chosenAlias) {
                        if (chosenAlias != null) {
                            try {
                                KeyChain.setKeyPair(getCurrentActivity(), chosenAlias, privateKey, certificates.toArray(new X509Certificate[certificates.size()]));
                                promise.resolve("Certificate installed successfully");
                            } catch (Exception e) {
                                Log.e(TAG, "Failed to set key pair", e);
                                promise.reject("Failed to set key pair", e);
                            }
                        } else {
                            promise.reject("User canceled", "User canceled the certificate installation");
                        }
                    }
                },
                null,
                null,
                null,
                -1,
                null
            );

        } catch (Exception e) {
            Log.e(TAG, "Failed to install certificate", e);
            promise.reject("Failed to install certificate", e);
        }
    }
}

File: android/app/src/main/java/com/pvarki/rasenmaeher/KeyChainModule.kt
package com.pvarki.rasenmaeher

import android.content.Intent
import android.security.KeyChain
import android.security.KeyChainAliasCallback
import android.util.Base64
import android.util.Log
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import java.security.KeyStore
import java.security.PrivateKey
import java.security.cert.Certificate
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate
import java.util.ArrayList

class KeyChainModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String {
        return "KeyChainModule"
    }

    @ReactMethod
    fun installCertificate(p12: String, name: String, promise: Promise) {
        try {
            val p12Bytes = Base64.decode(p12, Base64.DEFAULT)
            val keyStore = KeyStore.getInstance("PKCS12")
            keyStore.load(p12Bytes.inputStream(), "".toCharArray())

            val alias = keyStore.aliases().nextElement()
            val entry = keyStore.getEntry(alias, null) as KeyStore.PrivateKeyEntry

            val privateKey = entry.privateKey
            val certChain = entry.certificateChain
            val certificates = ArrayList<X509Certificate>()
            for (certificate in certChain) {
                certificates.add(certificate as X509Certificate)
            }

            KeyChain.choosePrivateKeyAlias(
                currentActivity,
                KeyChainAliasCallback { chosenAlias ->
                    if (chosenAlias != null) {
                        try {
                            KeyChain.setKeyPair(
                                currentActivity,
                                chosenAlias,
                                privateKey,
                                certificates.toTypedArray()
                            )
                            promise.resolve("Certificate installed successfully")
                        } catch (e: Exception) {
                            Log.e("KeyChainModule", "Failed to set key pair", e)
                            promise.reject("Failed to set key pair", e)
                        }
                    } else {
                        promise.reject("User canceled", "User canceled the certificate installation")
                    }
                },
                null,
                null,
                null,
                -1,
                null
            )

        } catch (e: Exception) {
            Log.e("KeyChainModule", "Failed to install certificate", e)
            promise.reject("Failed to install certificate", e)
        }
    }
}
